File: clevertap-geofence/src/test/java/com/clevertap/android/geofence/LocationUpdateTaskTest.java
package com.clevertap.android.geofence;

import static org.mockito.ArgumentMatchers.*;
import static org.mockito.Mockito.*;
import android.app.PendingIntent;
import android.content.Context;
import com.clevertap.android.geofence.interfaces.CTGeofenceTask;
import com.clevertap.android.geofence.interfaces.CTLocationAdapter;
import org.junit.*;
import org.mockito.*;

public class LocationUpdateTaskTest extends BaseTestCase {

    @Mock
    public CTGeofenceAPI ctGeofenceAPI;

    @Mock
    public CTLocationAdapter ctLocationAdapter;

    @Mock
    public CTGeofenceTask.OnCompleteListener onCompleteListener;

    @Mock
    private Logger logger;

    @Mock
    PendingIntent pendingIntent;

    private MockedStatic<CTGeofenceAPI> ctGeofenceAPIMockedStatic;

    private MockedStatic<FileUtils> fileUtilsMockedStatic;
    private MockedStatic<Utils> utilsMockedStatic;

    private MockedStatic<PendingIntentFactory> pendingIntentFactoryMockedStatic;

    @Before
    public void setUp() throws Exception {

        MockitoAnnotations.openMocks(this);

        ctGeofenceAPIMockedStatic = Mockito.mockStatic(CTGeofenceAPI.class);
        fileUtilsMockedStatic = Mockito.mockStatic(FileUtils.class);
        utilsMockedStatic = Mockito.mockStatic(Utils.class);
        pendingIntentFactoryMockedStatic = Mockito.mockStatic(PendingIntentFactory.class);
        super.setUp();

        when(CTGeofenceAPI.getInstance(application)).thenReturn(ctGeofenceAPI);
        when(CTGeofenceAPI.getLogger()).thenReturn(logger);

        when(ctGeofenceAPI.getCtLocationAdapter()).thenReturn(ctLocationAdapter);

    }

    @After
    public void cleanup() {
        ctGeofenceAPIMockedStatic.close();
        fileUtilsMockedStatic.close();
        utilsMockedStatic.close();
        pendingIntentFactoryMockedStatic.close();
    }

    @Test
    public void testExecuteTC1() {
        // when pending intent is null and bgLocationUpdate is true

        CTGeofenceSettings ctGeofenceSettings = new CTGeofenceSettings.Builder()
                .enableBackgroundLocationUpdates(true).build();
        when(ctGeofenceAPI.getGeofenceSettings()).thenReturn(ctGeofenceSettings);

        LocationUpdateTask task = new LocationUpdateTask(application);
        task.setOnCompleteListener(onCompleteListener);
        task.execute();

        verify(ctLocationAdapter).requestLocationUpdates();

        utilsMockedStatic.verify(() -> Utils.writeSettingsToFile(any(Context.class), any(CTGeofenceSettings.class)));

        verify(onCompleteListener).onComplete();
    }

    @Test
    public void testExecuteTC2() {
        // when pending intent is null and bgLocationUpdate is false

        CTGeofenceSettings ctGeofenceSettings = new CTGeofenceSettings.Builder()
                .enableBackgroundLocationUpdates(false).build();

        when(ctGeofenceAPI.getGeofenceSettings()).thenReturn(ctGeofenceSettings);

        LocationUpdateTask task = new LocationUpdateTask(application);
        task.execute();

        verify(ctLocationAdapter, never()).requestLocationUpdates();
        verify(ctLocationAdapter, never()).removeLocationUpdates(any(PendingIntent.class));

        utilsMockedStatic.verify(() -> Utils.writeSettingsToFile(any(Context.class), any(CTGeofenceSettings.class)));

    }

    @Test
    public void testExecuteTC3() {
        // when pending intent is not null and bgLocationUpdate is false

        CTGeofenceSettings ctGeofenceSettings = new CTGeofenceSettings.Builder()
                .enableBackgroundLocationUpdates(false).build();

        when(ctGeofenceAPI.getGeofenceSettings()).thenReturn(ctGeofenceSettings);

        // make pending intent non-null
        when(PendingIntentFactory.getPendingIntent(any(Context.class),any(int.class),any(int.class))).thenReturn(pendingIntent);

        LocationUpdateTask task = new LocationUpdateTask(application);
        task.execute();

        // verify that previous update is removed
        verify(ctLocationAdapter).removeLocationUpdates(any(PendingIntent.class));
        verify(ctLocationAdapter, never()).requestLocationUpdates();

        utilsMockedStatic.verify(() -> Utils.writeSettingsToFile(any(Context.class), any(CTGeofenceSettings.class)));

    }

    @Test
    public void testExecuteWhenLocationAdapterIsNull() {
        // when location adapter is null

        when(ctGeofenceAPI.getCtLocationAdapter()).thenReturn(null);
        LocationUpdateTask task = new LocationUpdateTask(application);
        task.execute();

        utilsMockedStatic.verifyNoInteractions();

    }

    @Test
    public void testIsRequestLocationTC1() {
        // when currentBgLocationUpdate is false
//
//        CTGeofenceSettings currentGeofenceSettings = new CTGeofenceSettings.Builder()
//                .enableBackgroundLocationUpdates(false).build();
//
//        when(ctGeofenceAPI.getGeofenceSettings()).thenReturn(currentGeofenceSettings);
//
//        CTGeofenceSettings lastGeofenceSettings = new CTGeofenceSettings.Builder().build();
//
//        LocationUpdateTask task = new LocationUpdateTask(application);
//        boolean isRequestLocation = WhiteboxImpl.invokeMethod(task, "isRequestLocation", null);
//
//        assertFalse(isRequestLocation);

    }

    @Test
    public void testIsRequestLocationTC10() {
        // when currentBgLocationUpdate is true and there is no change in settings

        CTGeofenceSettings currentGeofenceSettings = new CTGeofenceSettings.Builder()
                .enableBackgroundLocationUpdates(true)
                .build();

        when(ctGeofenceAPI.getGeofenceSettings()).thenReturn(currentGeofenceSettings);
        when(PendingIntentFactory.getPendingIntent(any(Context.class),any(int.class),any(int.class))).thenReturn(pendingIntent);

        CTGeofenceSettings lastGeofenceSettings = new CTGeofenceSettings.Builder()
                .build();

        when(Utils.readSettingsFromFile(application)).thenReturn(lastGeofenceSettings);

        LocationUpdateTask task = new LocationUpdateTask(application);
        task.execute();

        verify(logger).verbose(CTGeofenceAPI.GEOFENCE_LOG_TAG,
                "Dropping duplicate location update request");

    }

    @Test
    public void testIsRequestLocationTC2() {
        // when currentBgLocationUpdate is true and pendingIntent is null

        CTGeofenceSettings currentGeofenceSettings = new CTGeofenceSettings.Builder()
                .enableBackgroundLocationUpdates(true).build();

        when(PendingIntentFactory.getPendingIntent(any(Context.class),any(int.class),any(int.class))).thenReturn(null);
        when(ctGeofenceAPI.getGeofenceSettings()).thenReturn(currentGeofenceSettings);

        CTGeofenceSettings lastGeofenceSettings = new CTGeofenceSettings.Builder().build();

        when(Utils.readSettingsFromFile(application)).thenReturn(lastGeofenceSettings);

        LocationUpdateTask task = new LocationUpdateTask(application);
        task.execute();

        verify(ctLocationAdapter).requestLocationUpdates();

    }

    @Test
    public void testIsRequestLocationTC3() {
        // when fetch mode is current and change in accuracy

        CTGeofenceSettings currentGeofenceSettings = new CTGeofenceSettings.Builder()
                .enableBackgroundLocationUpdates(true)
                .setLocationFetchMode(CTGeofenceSettings.FETCH_CURRENT_LOCATION_PERIODIC)
                .setLocationAccuracy(CTGeofenceSettings.ACCURACY_HIGH)
                .build();

        when(ctGeofenceAPI.getGeofenceSettings()).thenReturn(currentGeofenceSettings);
        when(PendingIntentFactory.getPendingIntent(any(Context.class),any(int.class),any(int.class))).thenReturn(pendingIntent);


        CTGeofenceSettings lastGeofenceSettings = new CTGeofenceSettings.Builder()
                .setLocationAccuracy(CTGeofenceSettings.ACCURACY_LOW)
                .build();

        when(Utils.readSettingsFromFile(application)).thenReturn(lastGeofenceSettings);

        LocationUpdateTask task = new LocationUpdateTask(application);
        task.execute();

        verify(ctLocationAdapter).requestLocationUpdates();
    }

    @Test
    public void testIsRequestLocationTC4() {
        // when fetch mode is current and change in interval

        CTGeofenceSettings currentGeofenceSettings = new CTGeofenceSettings.Builder()
                .enableBackgroundLocationUpdates(true)
                .setLocationFetchMode(CTGeofenceSettings.FETCH_CURRENT_LOCATION_PERIODIC)
                .setInterval(2000000)
                .build();

        when(ctGeofenceAPI.getGeofenceSettings()).thenReturn(currentGeofenceSettings);
        when(PendingIntentFactory.getPendingIntent(any(Context.class),any(int.class),any(int.class))).thenReturn(pendingIntent);


        CTGeofenceSettings lastGeofenceSettings = new CTGeofenceSettings.Builder()
                .setInterval(5000000)
                .build();

        when(Utils.readSettingsFromFile(application)).thenReturn(lastGeofenceSettings);

        LocationUpdateTask task = new LocationUpdateTask(application);
        task.execute();

        verify(ctLocationAdapter).requestLocationUpdates();
    }

    @Test
    public void testIsRequestLocationTC5() {
        // when fetch mode is current and change in fastest interval

        CTGeofenceSettings currentGeofenceSettings = new CTGeofenceSettings.Builder()
                .enableBackgroundLocationUpdates(true)
                .setLocationFetchMode(CTGeofenceSettings.FETCH_CURRENT_LOCATION_PERIODIC)
                .setFastestInterval(2000000)
                .build();

        when(ctGeofenceAPI.getGeofenceSettings()).thenReturn(currentGeofenceSettings);
        when(PendingIntentFactory.getPendingIntent(any(Context.class),any(int.class),any(int.class))).thenReturn(pendingIntent);

        CTGeofenceSettings lastGeofenceSettings = new CTGeofenceSettings.Builder()
                .setFastestInterval(5000000)
                .build();

        when(Utils.readSettingsFromFile(application)).thenReturn(lastGeofenceSettings);

        LocationUpdateTask task = new LocationUpdateTask(application);
        task.execute();

        verify(ctLocationAdapter).requestLocationUpdates();

    }

    @Test
    public void testIsRequestLocationTC6() {
        // when fetch mode is current and change in displacement

        CTGeofenceSettings currentGeofenceSettings = new CTGeofenceSettings.Builder()
                .enableBackgroundLocationUpdates(true)
                .setLocationFetchMode(CTGeofenceSettings.FETCH_CURRENT_LOCATION_PERIODIC)
                .setSmallestDisplacement(600)
                .build();

        when(ctGeofenceAPI.getGeofenceSettings()).thenReturn(currentGeofenceSettings);
        when(PendingIntentFactory.getPendingIntent(any(Context.class),any(int.class),any(int.class))).thenReturn(pendingIntent);


        CTGeofenceSettings lastGeofenceSettings = new CTGeofenceSettings.Builder()
                .setSmallestDisplacement(700)
                .build();

        when(Utils.readSettingsFromFile(application)).thenReturn(lastGeofenceSettings);

        LocationUpdateTask task = new LocationUpdateTask(application);
        task.execute();

        verify(ctLocationAdapter).requestLocationUpdates();
    }

    @Test
    public void testIsRequestLocationTC7() {
        // when fetch mode is last location and change in interval

        CTGeofenceSettings currentGeofenceSettings = new CTGeofenceSettings.Builder()
                .enableBackgroundLocationUpdates(true)
                .setLocationFetchMode(CTGeofenceSettings.FETCH_LAST_LOCATION_PERIODIC)
                .setInterval(2000000)
                .build();

        when(ctGeofenceAPI.getGeofenceSettings()).thenReturn(currentGeofenceSettings);

        CTGeofenceSettings lastGeofenceSettings = new CTGeofenceSettings.Builder()
                .setInterval(5000000)
                .build();

        when(Utils.readSettingsFromFile(application)).thenReturn(lastGeofenceSettings);
        when(PendingIntentFactory.getPendingIntent(any(Context.class),any(int.class),any(int.class))).thenReturn(pendingIntent);

        LocationUpdateTask task = new LocationUpdateTask(application);
        task.execute();

        verify(ctLocationAdapter).requestLocationUpdates();
    }

    @Test
    public void testIsRequestLocationTC8() {
        // when fetch mode is current location and change in fetch mode

        CTGeofenceSettings currentGeofenceSettings = new CTGeofenceSettings.Builder()
                .enableBackgroundLocationUpdates(true)
                .setLocationFetchMode(CTGeofenceSettings.FETCH_CURRENT_LOCATION_PERIODIC)
                .build();

        when(ctGeofenceAPI.getGeofenceSettings()).thenReturn(currentGeofenceSettings);

        CTGeofenceSettings lastGeofenceSettings = new CTGeofenceSettings.Builder()
                .setLocationFetchMode(CTGeofenceSettings.FETCH_LAST_LOCATION_PERIODIC)
                .build();

        when(Utils.readSettingsFromFile(application)).thenReturn(lastGeofenceSettings);
        when(PendingIntentFactory.getPendingIntent(any(Context.class),any(int.class),any(int.class))).thenReturn(pendingIntent);

        LocationUpdateTask task = new LocationUpdateTask(application);
        task.execute();

        verify(ctLocationAdapter).requestLocationUpdates();
    }

    @Test
    public void testIsRequestLocationTC9() {
        // when fetch mode is last location and change in fetch mode

        CTGeofenceSettings currentGeofenceSettings = new CTGeofenceSettings.Builder()
                .enableBackgroundLocationUpdates(true)
                .setLocationFetchMode(CTGeofenceSettings.FETCH_LAST_LOCATION_PERIODIC)
                .build();

        when(ctGeofenceAPI.getGeofenceSettings()).thenReturn(currentGeofenceSettings);

        CTGeofenceSettings lastGeofenceSettings = new CTGeofenceSettings.Builder()
                .setLocationFetchMode(CTGeofenceSettings.FETCH_CURRENT_LOCATION_PERIODIC)
                .build();

        when(Utils.readSettingsFromFile(application)).thenReturn(lastGeofenceSettings);
        when(PendingIntentFactory.getPendingIntent(any(Context.class),any(int.class),any(int.class))).thenReturn(pendingIntent);


        LocationUpdateTask task = new LocationUpdateTask(application);
        task.execute();

        verify(ctLocationAdapter).requestLocationUpdates();
    }

}

File: clevertap-geofence/src/test/java/com/clevertap/android/geofence/LocationUpdateTaskTest.kt
package com.clevertap.android.geofence

import android.app.PendingIntent
import com.clevertap.android.geofence.interfaces.CTGeofenceTask
import com.clevertap.android.geofence.interfaces.CTLocationAdapter
import io.mockk.every
import io.mockk.mockk
import io.mockk.mockkStatic
import io.mockk.unmockkStatic
import io.mockk.verify
import org.junit.After
import org.junit.Before
import org.junit.Test

class LocationUpdateTaskTest : BaseTestCase() {

    private lateinit var ctGeofenceAPI: CTGeofenceAPI
    private lateinit var ctLocationAdapter: CTLocationAdapter
    private lateinit var onCompleteListener: CTGeofenceTask.OnCompleteListener
    private lateinit var logger: Logger
    private lateinit var pendingIntent: PendingIntent

    @Before
    override fun setUp() {
        super.setUp()

        ctGeofenceAPI = mockk(relaxed = true)
        ctLocationAdapter = mockk(relaxed = true)
        onCompleteListener = mockk(relaxed = true)
        logger = mockk(relaxed = true)
        pendingIntent = mockk(relaxed = true)

        mockkStatic(CTGeofenceAPI::class)
        mockkStatic(FileUtils::class)
        mockkStatic(Utils::class)
        mockkStatic(PendingIntentFactory::class)

        every { CTGeofenceAPI.getInstance(application) } returns ctGeofenceAPI
        every { CTGeofenceAPI.getLogger() } returns logger
        every { ctGeofenceAPI.ctLocationAdapter } returns ctLocationAdapter
    }

    @After
    override fun cleanUp() {
        super.cleanUp()
        unmockkStatic(CTGeofenceAPI::class)
        unmockkStatic(FileUtils::class)
        unmockkStatic(Utils::class)
        unmockkStatic(PendingIntentFactory::class)
    }

    @Test
    fun testExecuteTC1() {
        // when pending intent is null and bgLocationUpdate is true

        val ctGeofenceSettings =
            CTGeofenceSettings.Builder().enableBackgroundLocationUpdates(true).build()
        every { ctGeofenceAPI.geofenceSettings } returns ctGeofenceSettings

        val task = LocationUpdateTask(application)
        task.setOnCompleteListener(onCompleteListener)
        task.execute()

        verify { ctLocationAdapter.requestLocationUpdates() }

        verify { Utils.writeSettingsToFile(any(), any()) }

        verify { onCompleteListener.onComplete() }
    }

    @Test
    fun testExecuteTC2() {
        // when pending intent is null and bgLocationUpdate is false

        val ctGeofenceSettings =
            CTGeofenceSettings.Builder().enableBackgroundLocationUpdates(false).build()

        every { ctGeofenceAPI.geofenceSettings } returns ctGeofenceSettings

        val task = LocationUpdateTask(application)
        task.execute()

        verify(exactly = 0) { ctLocationAdapter.requestLocationUpdates() }
        verify(exactly = 0) { ctLocationAdapter.removeLocationUpdates(any()) }
        verify { Utils.writeSettingsToFile(any(), any()) }
    }

    @Test
    fun testExecuteTC3() {
        // when pending intent is not null and bgLocationUpdate is false

        val ctGeofenceSettings =
            CTGeofenceSettings.Builder().enableBackgroundLocationUpdates(false).build()

        every { ctGeofenceAPI.geofenceSettings } returns ctGeofenceSettings

        // make pending intent non-null
        every { PendingIntentFactory.getPendingIntent(any(), any(), any()) } returns pendingIntent

        val task = LocationUpdateTask(application)
        task.execute()

        // verify that previous update is removed
        verify { ctLocationAdapter.removeLocationUpdates(any()) }
        verify(exactly = 0) { ctLocationAdapter.requestLocationUpdates() }
        verify { Utils.writeSettingsToFile(any(), any()) }
    }

    @Test
    fun testExecuteWhenLocationAdapterIsNull() {
        // when location adapter is null

        every { ctGeofenceAPI.ctLocationAdapter } returns null
        val task = LocationUpdateTask(application)
        task.execute()

        verify(exactly = 0) { Utils.writeSettingsToFile(any(), any()) }
    }

    @Test
    fun testIsRequestLocationTC10() {
        // when currentBgLocationUpdate is true and there is no change in settings

        val currentGeofenceSettings =
            CTGeofenceSettings.Builder().enableBackgroundLocationUpdates(true).build()

        every { ctGeofenceAPI.geofenceSettings } returns currentGeofenceSettings
        every { PendingIntentFactory.getPendingIntent(any(), any(), any()) } returns pendingIntent

        val lastGeofenceSettings = CTGeofenceSettings.Builder().build()

        every { Utils.readSettingsFromFile(application) } returns lastGeofenceSettings

        val task = LocationUpdateTask(application)
        task.execute()

        verify {
            logger.verbose(
                CTGeofenceAPI.GEOFENCE_LOG_TAG, "Dropping duplicate location update request"
            )
        }
    }

    @Test
    fun testIsRequestLocationTC2() {
        // when currentBgLocationUpdate is true and pendingIntent is null

        val currentGeofenceSettings =
            CTGeofenceSettings.Builder().enableBackgroundLocationUpdates(true).build()

        every { PendingIntentFactory.getPendingIntent(any(), any(), any()) } returns null
        every { ctGeofenceAPI.geofenceSettings } returns currentGeofenceSettings

        val lastGeofenceSettings = CTGeofenceSettings.Builder().build()

        every { Utils.readSettingsFromFile(application) } returns lastGeofenceSettings

        val task = LocationUpdateTask(application)
        task.execute()

        verify { ctLocationAdapter.requestLocationUpdates() }
    }

    @Test
    fun testIsRequestLocationTC3() {
        // when fetch mode is current and change in accuracy

        val currentGeofenceSettings =
            CTGeofenceSettings.Builder().enableBackgroundLocationUpdates(true)
                .setLocationFetchMode(CTGeofenceSettings.FETCH_CURRENT_LOCATION_PERIODIC)
                .setLocationAccuracy(CTGeofenceSettings.ACCURACY_HIGH).build()

        every { ctGeofenceAPI.geofenceSettings } returns currentGeofenceSettings
        every { PendingIntentFactory.getPendingIntent(any(), any(), any()) } returns pendingIntent

        val lastGeofenceSettings =
            CTGeofenceSettings.Builder().setLocationAccuracy(CTGeofenceSettings.ACCURACY_LOW)
                .build()

        every { Utils.readSettingsFromFile(application) } returns lastGeofenceSettings

        val task = LocationUpdateTask(application)
        task.execute()

        verify { ctLocationAdapter.requestLocationUpdates() }
    }

    @Test
    fun testIsRequestLocationTC4() {
        // when fetch mode is current and change in interval

        val currentGeofenceSettings =
            CTGeofenceSettings.Builder().enableBackgroundLocationUpdates(true)
                .setLocationFetchMode(CTGeofenceSettings.FETCH_CURRENT_LOCATION_PERIODIC)
                .setInterval(2000000).build()

        every { ctGeofenceAPI.geofenceSettings } returns currentGeofenceSettings
        every { PendingIntentFactory.getPendingIntent(any(), any(), any()) } returns pendingIntent


        val lastGeofenceSettings = CTGeofenceSettings.Builder().setInterval(5000000).build()
        every { Utils.readSettingsFromFile(application) } returns lastGeofenceSettings

        val task = LocationUpdateTask(application)
        task.execute()

        verify { ctLocationAdapter.requestLocationUpdates() }
    }

    @Test
    fun testIsRequestLocationTC5() {
        // when fetch mode is current and change in fastest interval

        val currentGeofenceSettings =
            CTGeofenceSettings.Builder().enableBackgroundLocationUpdates(true)
                .setLocationFetchMode(CTGeofenceSettings.FETCH_CURRENT_LOCATION_PERIODIC)
                .setFastestInterval(2000000).build()

        every { ctGeofenceAPI.geofenceSettings } returns currentGeofenceSettings
        every { PendingIntentFactory.getPendingIntent(any(), any(), any()) } returns pendingIntent

        val lastGeofenceSettings = CTGeofenceSettings.Builder().setFastestInterval(5000000).build()

        every { Utils.readSettingsFromFile(application) } returns lastGeofenceSettings

        val task = LocationUpdateTask(application)
        task.execute()

        verify { ctLocationAdapter.requestLocationUpdates() }
    }

    @Test
    fun testIsRequestLocationTC6() {
        // when fetch mode is current and change in displacement

        val currentGeofenceSettings =
            CTGeofenceSettings.Builder().enableBackgroundLocationUpdates(true)
                .setLocationFetchMode(CTGeofenceSettings.FETCH_CURRENT_LOCATION_PERIODIC)
                .setSmallestDisplacement(600f).build()

        every { ctGeofenceAPI.geofenceSettings } returns currentGeofenceSettings
        every { PendingIntentFactory.getPendingIntent(any(), any(), any()) } returns pendingIntent

        val lastGeofenceSettings =
            CTGeofenceSettings.Builder().setSmallestDisplacement(700f).build()

        every { Utils.readSettingsFromFile(application) } returns lastGeofenceSettings

        val task = LocationUpdateTask(application)
        task.execute()

        verify { ctLocationAdapter.requestLocationUpdates() }
    }

    @Test
    fun testIsRequestLocationTC7() {
        // when fetch mode is last location and change in interval

        val currentGeofenceSettings =
            CTGeofenceSettings.Builder().enableBackgroundLocationUpdates(true)
                .setLocationFetchMode(CTGeofenceSettings.FETCH_LAST_LOCATION_PERIODIC)
                .setInterval(2000000).build()

        every { ctGeofenceAPI.geofenceSettings } returns currentGeofenceSettings

        val lastGeofenceSettings = CTGeofenceSettings.Builder().setInterval(5000000).build()

        every { Utils.readSettingsFromFile(application) } returns lastGeofenceSettings
        every { PendingIntentFactory.getPendingIntent(any(), any(), any()) } returns pendingIntent

        val task = LocationUpdateTask(application)
        task.execute()

        verify { ctLocationAdapter.requestLocationUpdates() }
    }

    @Test
    fun testIsRequestLocationTC8() {
        // when fetch mode is current location and change in fetch mode

        val currentGeofenceSettings =
            CTGeofenceSettings.Builder().enableBackgroundLocationUpdates(true)
                .setLocationFetchMode(CTGeofenceSettings.FETCH_CURRENT_LOCATION_PERIODIC).build()

        every { ctGeofenceAPI.geofenceSettings } returns currentGeofenceSettings

        val lastGeofenceSettings = CTGeofenceSettings.Builder()
            .setLocationFetchMode(CTGeofenceSettings.FETCH_LAST_LOCATION_PERIODIC).build()

        every { Utils.readSettingsFromFile(application) } returns lastGeofenceSettings
        every { PendingIntentFactory.getPendingIntent(any(), any(), any()) } returns pendingIntent

        val task = LocationUpdateTask(application)
        task.execute()

        verify { ctLocationAdapter.requestLocationUpdates() }
    }

    @Test
    fun testIsRequestLocationTC9() {
        // when fetch mode is last location and change in fetch mode

        val currentGeofenceSettings =
            CTGeofenceSettings.Builder().enableBackgroundLocationUpdates(true)
                .setLocationFetchMode(CTGeofenceSettings.FETCH_LAST_LOCATION_PERIODIC).build()

        every { ctGeofenceAPI.geofenceSettings } returns currentGeofenceSettings

        val lastGeofenceSettings = CTGeofenceSettings.Builder()
            .setLocationFetchMode(CTGeofenceSettings.FETCH_CURRENT_LOCATION_PERIODIC).build()

        every { Utils.readSettingsFromFile(application) } returns lastGeofenceSettings
        every { PendingIntentFactory.getPendingIntent(any(), any(), any()) } returns pendingIntent

        val task = LocationUpdateTask(application)
        task.execute()

        verify { ctLocationAdapter.requestLocationUpdates() }
    }
}
